ethereum/blockwatch/block_watcher.go

```diff
@@ -302,14 +306,24 @@ func (w *Watcher) SyncToLatestBlock() error {
 		return syncErr
 	}
 	if w.shouldRevertChanges(lastStoredHeader, allEvents) {
-		// TODO(albrow): Take another look at this. Maybe extract to method.
-		if err := w.stack.Clear(); err != nil {
+		if err := w.stack.Reset(checkpoint); err != nil {
+			return err
+		}
+	} else {
+		_, err := w.stack.Checkpoint()
+		if err != nil {
 			return err
 		}
-		if _, _, err := w.stack.db.AddMiniHeaders(existingMiniHeaders); err != nil {
+		if _, err := w.db.DeleteMiniHeaders(nil); err != nil {
```

Comment:
We need to do `DeleteMiniHeaders` and `AddMiniHeaders` in a transaction so that the database is always in a good state. Otherwise orderwatcher could read from the database in between when we call `DeleteMiniHeaders` and when we call `AddMiniHeaders`.

The current version of the `db` package does not expose a way to execute arbitrary transactions, partly for separation of concerns and partly due to the difficulty of designing a transaction interface that works with both SQL and Dexie.js. So I think our best option is to add a new method to the `db` package called something like `ResetMiniHeaders` which uses a transaction under the hood.